File: app/src/main/java/com/krystianwsul/checkme/gui/instances/tree/UnscheduledNode.java
package com.krystianwsul.checkme.gui.instances.tree;

import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.View;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.gui.tree.ModelNode;
import com.krystianwsul.checkme.gui.tree.NodeContainer;
import com.krystianwsul.checkme.gui.tree.TreeNode;
import com.krystianwsul.checkme.utils.TaskKey;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.List;

class UnscheduledNode extends GroupHolderNode implements ModelNode, TaskParent {
    @NonNull
    private final NodeCollection mNodeCollection;

    private TreeNode mTreeNode;

    private List<TaskNode> mTaskNodes;

    UnscheduledNode(float density, @NonNull NodeCollection nodeCollection) {
        super(density, 0);

        mNodeCollection = nodeCollection;
    }

    @NonNull
    TreeNode initialize(boolean expanded, @NonNull NodeContainer nodeContainer, @NonNull List<GroupListFragment.TaskData> taskDatas, @Nullable List<TaskKey> expandedTaskKeys) {
        Assert.assertTrue(!expanded || !taskDatas.isEmpty());

        mTreeNode = new TreeNode(this, nodeContainer, expanded, false);

        mTaskNodes = new ArrayList<>();

        List<TreeNode> childTreeNodes = Stream.of(taskDatas)
                .map(taskData -> newChildTreeNode(taskData, expandedTaskKeys))
                .collect(Collectors.toList());

        mTreeNode.setChildTreeNodes(childTreeNodes);

        return mTreeNode;
    }

    @NonNull
    private TreeNode newChildTreeNode(@NonNull GroupListFragment.TaskData taskData, @Nullable List<TaskKey> expandedTaskKeys) {
        TaskNode taskNode = new TaskNode(mDensity, 0, taskData, this);

        mTaskNodes.add(taskNode);

        return taskNode.initialize(getTreeNode(), expandedTaskKeys);
    }

    @NonNull
    private NodeCollection getNodeCollection() {
        return mNodeCollection;
    }

    boolean expanded() {
        return getTreeNode().expanded();
    }

    List<TaskKey> getExpandedTaskKeys() {
        return Stream.of(mTaskNodes)
                .flatMap(TaskNode::getExpandedTaskKeys)
                .collect(Collectors.toList());
    }

    @NonNull
    @Override
    public GroupListFragment.GroupAdapter getGroupAdapter() {
        return getNodeCollection().getGroupAdapter();
    }

    @NonNull
    private GroupListFragment getGroupListFragment() {
        return getGroupAdapter().mGroupListFragment;
    }

    @NonNull
    private TreeNode getTreeNode() {
        Assert.assertTrue(mTreeNode != null);

        return mTreeNode;
    }

    @Override
    public int compareTo(@NonNull ModelNode modelNode) {
        if (modelNode instanceof DividerNode) {
            return -1;
        } else {
            Assert.assertTrue(modelNode instanceof NotDoneGroupNode);
            return 1;
        }
    }

    @Override
    int getNameVisibility() {
        return View.VISIBLE;
    }

    @NonNull
    @Override
    String getName() {
        return getGroupListFragment().getString(R.string.noReminder);
    }

    @Override
    int getNameColor() {
        return ContextCompat.getColor(getGroupListFragment().getActivity(), R.color.textPrimary);
    }

    @Override
    boolean getNameSingleLine() {
        return true;
    }

    @Override
    int getDetailsVisibility() {
        return View.GONE;
    }

    @NonNull
    @Override
    String getDetails() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getDetailsColor() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getChildrenVisibility() {
        return View.GONE;
    }

    @NonNull
    @Override
    String getChildren() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getChildrenColor() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getExpandVisibility() {
        return View.VISIBLE;
    }

    @Override
    int getExpandImageResource() {
        if (getTreeNode().expanded())
            return R.drawable.ic_expand_less_black_36dp;
        else
            return R.drawable.ic_expand_more_black_36dp;
    }

    @NonNull
    @Override
    View.OnClickListener getExpandOnClickListener() {
        return getTreeNode().getExpandListener();
    }

    @Override
    int getCheckBoxVisibility() {
        return View.INVISIBLE;
    }

    @Override
    boolean getCheckBoxChecked() {
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    View.OnClickListener getCheckBoxOnClickListener() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getSeparatorVisibility() {
        return (getTreeNode().getSeparatorVisibility() ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    int getBackgroundColor() {
        return Color.TRANSPARENT;
    }

    @Override
    View.OnLongClickListener getOnLongClickListener() {
        return getTreeNode().getOnLongClickListener();
    }

    @Override
    View.OnClickListener getOnClickListener() {
        return getTreeNode().getOnClickListener();
    }

    @Override
    public boolean selectable() {
        return false;
    }

    @Override
    public void onClick() {

    }

    @Override
    public boolean visibleWhenEmpty() {
        return false;
    }

    @Override
    public boolean visibleDuringActionMode() {
        return false;
    }

    @Override
    public boolean separatorVisibleWhenNotExpanded() {
        return false;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/instances/tree/UnscheduledNode.kt
package com.krystianwsul.checkme.gui.instances.tree

import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.gui.instances.list.GroupListDataWrapper
import com.krystianwsul.checkme.gui.tasks.ShowTasksActivity
import com.krystianwsul.checkme.gui.tree.AbstractHolder
import com.krystianwsul.checkme.gui.tree.AbstractModelNode
import com.krystianwsul.checkme.gui.tree.DetailsNode
import com.krystianwsul.checkme.gui.tree.HolderType
import com.krystianwsul.checkme.gui.tree.delegates.expandable.ExpandableDelegate
import com.krystianwsul.checkme.gui.tree.delegates.indentation.IndentationDelegate
import com.krystianwsul.checkme.gui.tree.delegates.indentation.IndentationModelNode
import com.krystianwsul.checkme.gui.tree.delegates.singleline.SingleLineDelegate
import com.krystianwsul.checkme.gui.tree.delegates.singleline.SingleLineModelNode
import com.krystianwsul.checkme.gui.utils.flatten
import com.krystianwsul.common.utils.ProjectKey
import com.krystianwsul.common.utils.TaskKey
import com.krystianwsul.treeadapter.ModelNode
import com.krystianwsul.treeadapter.NodeContainer
import com.krystianwsul.treeadapter.TreeNode

class UnscheduledNode(
    private val nodeCollection: NodeCollection,
    private val unscheduledFirst: Boolean,
    private val projectKey: ProjectKey.Shared?,
) : AbstractModelNode(), TaskParent, SingleLineModelNode, IndentationModelNode {

    override val holderType = HolderType.EXPANDABLE_SINGLELINE

    override val id get() = Id(nodeCollection.nodeContainer.id)

    data class Id(val id: Any)

    private lateinit var taskDatas: List<GroupListDataWrapper.TaskData>

    override lateinit var treeNode: TreeNode<AbstractHolder>
        private set

    private val taskNodes = mutableListOf<TaskNode>()

    val taskExpansionStates get() = taskNodes.map { it.taskExpansionStates }.flatten()

    private val groupListFragment by lazy { groupAdapter.groupListFragment }

    override val parentNode: ModelNode<AbstractHolder>? = null

    override val indentation = 0

    override val delegates by lazy {
        listOf(
            ExpandableDelegate(treeNode),
            SingleLineDelegate(this),
            IndentationDelegate(this)
        )
    }

    fun initialize(
        expansionState: TreeNode.ExpansionState?,
        nodeContainer: NodeContainer<AbstractHolder>,
        taskDatas: List<GroupListDataWrapper.TaskData>,
        taskExpansionStates: Map<TaskKey, TreeNode.ExpansionState>,
        selectedTaskKeys: List<TaskKey>,
    ): TreeNode<AbstractHolder> {
        this.taskDatas = taskDatas

        treeNode = TreeNode(
            this,
            nodeContainer,
            initialExpansionState = expansionState.takeIf { taskDatas.isNotEmpty() },
        )

        treeNode.setChildTreeNodes(taskDatas.map { newChildTreeNode(it, taskExpansionStates, selectedTaskKeys) })

        return treeNode
    }

    private fun newChildTreeNode(
        taskData: GroupListDataWrapper.TaskData,
        taskExpansionStates: Map<TaskKey, TreeNode.ExpansionState>,
        selectedTaskKeys: List<TaskKey>,
    ) = TaskNode(0, taskData, this, this).let {
        taskNodes.add(it)

        it.initialize(treeNode, taskExpansionStates, selectedTaskKeys)
    }

    val expansionState get() = treeNode.expansionState

    override val groupAdapter by lazy { nodeCollection.groupAdapter }

    override fun onClick(holder: AbstractHolder) = groupListFragment.activity.startActivity(
        ShowTasksActivity.newIntent(ShowTasksActivity.Parameters.Unscheduled(projectKey))
    )

    override fun compareTo(other: ModelNode<AbstractHolder>) = when {
        unscheduledFirst -> -1
        other is DetailsNode -> -1
        other is DividerNode -> -1
        else -> {
            check(other is NotDoneGroupNode)

            1
        }
    }

    override val text by lazy { groupListFragment.activity.getString(R.string.notes) }


    override fun isVisible(actionMode: Boolean, hasVisibleChildren: Boolean): Boolean {
        return hasVisibleChildren
    }
}
